tests/kernels/test_self_and_cross_attn.py

```diff
@@ -0,0 +1,1711 @@
+
+    * Construct fake test vectors for self- and cross-attention
+    * Construct attention metadata structure with self- and cross-attention
+      attributes
+    * Test self- and cross-attention in the following order
+    
+        * Prefill self-attention
+        * Prefill cross-attention
+        * Decode self-attention
+        * Decode cross-attention
+        * This order would exacerbate any accidental overlap in the
+          self-/cross-attention block tables, which we attempt to avoid
+    * Validate output correctness against ideal reference attention
+      implementation
+
+    Block tables are constructed such that cross-attention KV cache is in a
+    higher, non-intersecting address-space than self-attention KV cache.
+
+    Self- and cross-attention share the same query tensor but not the K/V
+    tensors. Self-attention K/Vs must have the same seq len as Q while
+    cross-attention K/Vs are allowed to differ in seq len, as is often the case
+    for cross-attention.
+    '''
+
+    with backend_override_fixture(backend_name):
+        # Force Attention wrapper backend
+
+        # Num KV cache blocks
+        num_blocks = 4096
+
+        # Attention scale factor, attention backend instance, attention wrapper
+        # instance, KV cache init
+        scale, \
+        attn_backend, \
+        attn, \
+        kv_cache = basic_setup(num_heads,
+                              head_size,
+                              num_blocks,
+                              block_size,
+                              backend_name)
+
+        # Self-attention setup
+
+        self_block_base_addr = 0
+
+        query, \
+        prefill_packed_query, \
+        self_prefill_packed_key, \
+        self_prefill_packed_value, \
+        self_prefill_packed_ideal_output, \
+        prefill_q_seq_lens, \
+        self_prefill_kv_seq_lens, \
+        decode_packed_query, \
+        self_decode_packed_key, \
+        self_decode_packed_value, \
+        self_decode_packed_ideal_output, \
+        _, \
+        _, \
+        q_seq_lens, \
+        _, \
+        self_decode_block_tables, \
+        self_decode_slot_mapping, \
+        self_prefill_slot_mapping, \
+        self_prefill_block_tables, \
+        cross_block_base_addr = decoder_attn_setup(batch_size,
+                                                    num_heads,
+                                                    head_size,
+                                                    block_size,
+                                                    scale,
+                                                    max_q_seq_len,
+                                                    block_base_addr=self_block_base_addr)
+
+        # Cross-attention setup
+
+        cross_prefill_packed_key, \
+        cross_prefill_packed_value, \
+        cross_prefill_packed_ideal_output, \
+        cross_decode_packed_ideal_output, \
+        cross_kv_seq_lens, \
+        cross_decode_block_tables, \
```

Comment:
wondering if we need all 4  cross_decode_block_tables, cross_decode_slot_mapping, cross_prefill_slot_mapping and  cross_prefill_block_tables here? Won't  cross_prefill_slot_mapping and cross_decode_block_tables be sufficient - the first one for populating the kv cache and the other for the cross attention computation with decoder attention?

Reply:
This is a good question. I think you are alluding to the fact that cross_decode_slot_mapping and cross_decode_block_tables are empty tensors, and so can be omitted. While these tensors are empty, nonetheless they are *tensors* as opposed to being None, thus they still have to be constructed and passed as arguments to attention.

I agree that the test code is complicated by the large numbers of arguments and return-values in each function call.

I have remedied this by creating a set of named tuple data structures that encapsulate the arguments and return values.